docs/source/advanced/link-prediction.rst

```diff
@@ -0,0 +1,206 @@
+
+    * ``inbatch_joint``: In-batch joint negative sampling.
+
+.. _speedup_lp_training_label:
+
+Speedup Link Prediction Training
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+GraphStorm relies on ``dgl.dataloading.MultiLayerNeighborSampler`` and
+``train_mask`` to avoid sampling validation and test edges during training.
+Basically, it only samples edges with ``train_mask`` set to be `True`. However,
+the implementation is not efficient. To speedup graph sampling during link
+prediction training, GraphStorm provides four link prediction dataloaders
+(i.e., ``fast_uniform``, ``fast_joint``, ``fast_localuniform`` and
+``fast_localjoint``) with more efficient implementation but less precise
+neighbor sampling behavior. To be more specific, these dataloaders will do
+neighbor sampling regardless of any masks in the beginning, and later remove
+edges with  ``val_mask`` or ``test_mask`` set to be `True`. In theory, a sampled
+subgraph may have less neighbor nodes than expected as some of them would be
+removed. However, with a graph having hundreds of millions of edges (or more)
+and small validation and test sets, e.g., each with less than 10% edges, the
+impact is negligible.
+
+With DGL 1.0.4, ``fast_localuniform`` dataloader can speedup 2.4X over ``localuniform`` dataloader on training a 2 layer RGCN on MAG dataset on four g5.48x instances.
+
+Hard Negative sampling
+-----------------------
+GraphStorm provides support for users to define hard negative edges for a positive edge during Link Prediction training.
+Currently, hard negative edges are constructed by replacing the destination nodes of edges with pre-defined hard negatives.
+For example, given an edge (``src_pos``, ``dst_pos``) and its hard negative destination nodes ``hard_0`` and ``hard_1``, GraphStorm will construct two hard negative edges, i.e., (``src_pos``, ``hard_0``) and (``src_pos``, ``hard_1``).
+
+The hard negatives are stored as edge features of the target edge type.
+Users can provide the hard negatives for each edge type through ``train_etypes_negative_dstnode`` in the training config yaml.
+For example, the following yaml block defines the hard negatives for edge type (``src_type``,``rel_type0``,``dst_type``) as the edge feature ``negative_nid_field_0`` and the hard negatives for edge type ``(src_type,rel_type1,dst_type)`` as the edge feature ``negative_nid_field_1``.
+
+  .. code-block:: yaml
+
+    train_etypes_negative_dstnode:
+      - src_type,rel_type0,dst_type:negative_nid_field_0
+      - src_type,rel_type1,dst_type:negative_nid_field_1
+
+Users can also define the number of hard negatives to sample for each edge type during training though ``num_train_hard_negatives`` in the training config yaml.
+For example, the following yaml block defines the number of hard negatives for edge type (``src_type``,``rel_type0``,``dst_type``) is 5 and the number of hard negatives for edge type (``src_type``,``rel_type1``,``dst_type``) is 10.
```

Comment:
Sorry, the comment should be in this line, i.e., 

"(``src_type``,``rel_type0``,``dst_type``)" -> "(``src_type, rel_type0, 
dst_type``)"  and
"(``src_type``,``rel_type1``,``dst_type``)" -> "(``src_type, rel_type1, 
dst_type``)"